### ISSUE TITLE
Import always recreates catalog indices

### ISSUE THREAD

**Issue by tseaver** [zopefoundation/Products.GenericSetup#1]
In https://bugs.launchpad.net/zope-cmf/+bug/161682, @mauritsvanrees reported:

> At Zest we have an own Plone product feedfeeder for a client (soon to be released to the collective I think). We make use of GS. Its profiles/default/catalog.xml is:

```xml
<?xml version="1.0"?>
<object name="portal_catalog" meta_type="Pluggable Plone Catalog Tool">
 <index name="getFeedItemAuthor" meta_type="FieldIndex">
  <indexed_attr value="getFeedItemAuthor"/>
 </index>
 <index name="getFeedItemUpdated" meta_type="DateIndex">
  <property name="index_naive_time_as_local">True</property>
 </index>
 <column value="getFeedItemAuthor"/>
 <column value="getFeedItemUpdated"/>
</object>
```

> So we are adding some indexes. But on a reinstall those indexes get thrown away and readded. This means we need to reindex those indexes manually (or rather in code) so the catalog is up to date again.

> We tried to get GS to leave those indexes alone if they are already there, but failed. It looks like a bug in GS, but we could also do something wrong of course.

> We have a separate product with an Extentions/install.py which takes care of loading all the GS profiles of the products that we need and that handles some other install functions via the quickinstaller. The relevant part of the code:

```python
    setup_tool = getToolByName(portal, 'portal_setup')
    for extension_id in config.GS_DEPS:
        setup_tool.setImportContext('profile-%s' % extension_id)
        setup_tool.runAllImportSteps(purge_old=False)
```

> And the culprit is the purge_old in the last line. If we have followed the trail into GS right, this _should_ make sure that GS does not remove those indexes, but in fact it does. The indexes are there in the end, but none of the brains in the catalog have any content for those fields. Reindexing in some other part of the code solves it:

```python
    site.portal_catalog.manage_reindexIndex(ids=['getFeedItemAuthor','getFeedItemUpdated'])
```

> Would this be a GS bug or are we doing something wrong?

Yvo Schubbe replied:

> Not sure if this is a bug report or a feature request or if the bug is the fact the current limitations are not documented. Anyway it is a known issue.

> I quote some suggestions I made on the mailing list:
> ( http://mail.zope.org/pipermail/zope-cmf/2006-August/024747.html )

> - The adapters should compare the new settings with the existing ones and make sure indexes are only updated if they are actually changed. This is not implemented so far.

> - The adapters should modify existing indexes if the profile contains new settings. This is not implemented for the 'extra' elements.

> - Reindexing might be very expensive. That's the reason why GenericSetup does no reindexing, you have to do it manually. We need a better solution for that, maybe some bookkeeping which indexes need to be reindexed and a button in the setup tool that triggers reindexing.

@mauritsvanrees responded:

> I was not sure either when I filed this. :) Some time has passed, I asked some questions and learned some more. I would now say it is a feature request.

> I reported some ideas in this thread:

> http://article.gmane.org/gmane.comp.web.zope.cmf/13371

@tseaver wrote:

> Yuppie thinks he can see how to get the index adapters in GS to avoid clearing an index if the import does not change its properties.

@wichert wrote:

> Unfortunately there is no API to either get or update the values of the extra index parameters. That suggests that if extra is specified in the profile the index needs to be recreated.

@stainsby wrote:

> I recently reported this in the Plone tracker and it was suggested that I add a comment here instead. Here is my original report:

> "Catalog indexes that are defined by a product (in the catalog.xml file of its profile) are cleared each time a product is reinstalled. In fact, just deleting the product directory and re-adding also seems to do the same thing. With a client's membership database that takes one or two days to reindex, this is a major pain."

@dnouri posted two patches:

https://bugs.launchpad.net/zope-cmf/+bug/161682/+attachment/169881/+files/CMFCore-2.1-branch.diff

https://bugs.launchpad.net/zope-cmf/+bug/161682/+attachment/169882/+files/GenericSetup-1.3-branch.diff

with comment:

> Attached are two patches. One against GenericSetup in its 1.3 branch and one against CMFCore in the 2.1 branch.

> The goal of this patch is to work towards a solution for one part of this issue, namely, citing yuppie: "The adapters should compare the new settings with the existing ones and make sure indexes are only updated if they are actually changed."

> For this purpose I added a generally usable function decorator called `clearIfChanged` that you can stick onto existing `_importNode` methods. This decorator will compare self.node with the value of self.node *after* the node was imported. It does so to detect whether anything has changed; if anything has changed, it will call `self.context.clear()`.

> I added the tests for this to CMFCore; I must admit I was a bit confused as to where to put them, because CMFCore seems to contain in `CMFCore/exportimport/tests/test_catalog.py` tests that should maybe live in GenericSetup. I'm happy to move tests around if need be.

> Note that the only two indexes that I added this decorator to right now is the DateIndex and the ZCTextIndex. I'll add it to more index adapters once we have consensus that this approach is good, and everything is going into the right place.

@wichert responded:

> Looking at that patch it relies on being able to make a reliable XML export of the node so it can compare that with the new state. That is an assumption that you can not make though: only ZCTextIndex has an exporter that exports the extra options correctly. For all other index types this is not implemented. Until that has changed this patch can break sites.

@dnouri replied:

> If the export doesn't deal with the same data as the import, then that's a bug in the current handlers, and one that's quite easily fixed, I believe.

> I see no need to do the "interface change for catalog indexes in general" that you suggest in an e-mail. Since handlers are index specific, it's fine for them to introspect states of their indexes down to private attribute level.

> I'll try and see if I can get all indexes that I'm interested in to export properly. In the meantime, please help me understand how this patch can break sites.

Yvo Schubbe wrote:

> Thanks for the patch! 2 comments:

> - The export/import adapters are currently cheating extensively. There is no way to write them using just the interface methods, but they claim to adapt the interfaces. Things don't become worse if we use non-API methods for extracting and updating the 'extra' settings - in fact we adapt a specific implementation.

> - The tests in CMFCore/exportimport/tests/test_catalog.py are just additional integration tests for the CMF specific catalog tool. GenericSetup contains test_exportimport.py files for the indexes and ZCatalog. All these adapter tests in GS should be modified to test the new behavior.

Toni Mueller (toni.mueller@oeko.net) wrote:

> It would great to see some progress in this area. We're currently bitten by this bug as well, and re-indexing would take several hours each time. Before solving the problem automatically, I'd like to have just an extension to the syntax of catalog.xml:

> Now we have:

```xml
<indexed_attr value="some_attribute"/>
```

> It would imho be a big step forward if we had this instead:

```xml
<indexed_attr value="some_attribute" [preserve=True/False] [reindex=True/False] />
```

> With those two optional parameters, the decision about how to proceed with indices would not have to be computed, but could simply be specified by the user.

@rochecompaan followed up:

> I just hit this bug again today. I don't understand why the PluggableIndexNodeAdapter needs to ever clear the index in _importNode - it makes catalog.xml dangerous and unusable.

> Actually not only PluggableIndexNodeAdapter but all the other IndexNodeAdapter in Products/Generic Setup/PluginIndexes/exportimport.py as well.

@djay wrote:

> This feature still isn't implemented :(
> I might be missing something but does the solution have to involve changing the index api?
> Can't we just define the behaviour such that
>  - purge="True", the index will always get recreated.
>  - purge="False",if the index id already exists do nothing. if it doesn't then create it. If the extra attributes have changed bad luck.

**Comment by hvelarde** [comment]
I agree with @djay on this, why we can just add the `purge` directive defaulting to `False`?
